File: day5/main.ts
import { readFileSync } from "fs";
import { performance } from "perf_hooks";

type Range = [bigint, bigint];

// Merge overlapping ranges using reduce (idiomatic TypeScript)
function mergeRanges(ranges: Range[]): Range[] {
  const sorted = [...ranges].sort((a, b) =>
    a[0] === b[0] ? (a[1] < b[1] ? -1 : 1) : a[0] < b[0] ? -1 : 1,
  );

  return sorted.reduce<Range[]>((merged, [a, b]) => {
    if (merged.length === 0 || a > merged[merged.length - 1][1] + 1n) {
      merged.push([a, b]);
    } else if (b > merged[merged.length - 1][1]) {
      merged[merged.length - 1][1] = b;
    }
    return merged;
  }, []);
}

function inAny(ranges: Range[], x: bigint): boolean {
  let lo = 0;
  let hi = ranges.length;
  while (lo < hi) {
    const mid = Math.floor((lo + hi) / 2);
    if (ranges[mid][0] <= x) {
      lo = mid + 1;
    } else {
      hi = mid;
    }
  }
  if (lo === 0) return false;
  const [a, b] = ranges[lo - 1];
  return a <= x && x <= b;
}

function parse(text: string): [Range[], bigint[]] {
  const norm = text.replace(/\r\n/g, "\n");
  const parts = norm.split(/\n\n/);
  const top = parts[0] ?? "";
  const bottom = parts[1] ?? "";
  const ranges: Range[] = top
    .trim()
    .split(/\r?\n/)
    .filter((line) => line.trim())
    .map((line) => {
      const [a, b] = line.split("-").map((s) => BigInt(s));
      return [a, b] as Range;
    });
  const ids = bottom
    .trim()
    .split(/\r?\n/)
    .filter(Boolean)
    .map((s) => BigInt(s.trim()));
  return [ranges, ids];
}

function solve(text: string): [bigint, bigint] {
  const [ranges, ids] = parse(text);
  const merged = mergeRanges(ranges);

  // Part 1: Count IDs in any interval
  const fresh = ids.filter((id) => inAny(merged, id)).length;

  // Part 2: Total span using reduce
  const total = merged.reduce((sum, [a, b]) => sum + (b - a + 1n), 0n);

  return [BigInt(fresh), total];
}

const t0 = performance.now();
const text = readFileSync("input.txt", "utf8");
const [p1, p2] = solve(text);
const elapsed = performance.now() - t0;
console.log(
  `available_fresh=${p1} total_fresh_ids=${p2} elapsed_ms=${elapsed.toFixed(3)}`,
);

File: day5/main.rs
use std::collections::BTreeMap;
use std::fs;
use std::time::Instant;

fn parse(path: &str) -> (Vec<(i64, i64)>, Vec<i64>) {
    let raw = fs::read_to_string(path).expect("input");
    let text = raw.replace("\r\n", "\n");
    let mut parts = text.splitn(2, "\n\n");
    let top = parts.next().unwrap_or("");
    let bottom = parts.next().unwrap_or("");
    let ranges = top
        .lines()
        .filter_map(|line| {
            let line = line.trim();
            if line.is_empty() {
                None
            } else {
                let bits: Vec<_> = line.split('-').collect();
                Some((bits[0].parse().unwrap(), bits[1].parse().unwrap()))
            }
        })
        .collect();
    let ids = bottom
        .lines()
        .filter_map(|line| {
            let line = line.trim();
            if line.is_empty() {
                None
            } else {
                Some(line.parse().unwrap())
            }
        })
        .collect();
    (ranges, ids)
}

// Build a BTreeMap of merged intervals: start -> end
fn build_interval_map(ranges: Vec<(i64, i64)>) -> BTreeMap<i64, i64> {
    let mut sorted = ranges;
    sorted.sort_by_key(|r| (r.0, r.1));

    let mut map: BTreeMap<i64, i64> = BTreeMap::new();
    for (a, b) in sorted {
        // Find if we can merge with existing intervals
        // Get the interval that starts at or before 'a'
        let mut merged = false;
        if let Some((&start, &end)) = map.range(..=a).next_back() {
            if a <= end + 1 {
                // Can merge: extend the existing interval
                let new_end = end.max(b);
                map.insert(start, new_end);
                merged = true;
            }
        }
        if !merged {
            map.insert(a, b);
        }
    }
    map
}

// Check if a point is contained in any interval using BTreeMap range queries
fn contains_point(map: &BTreeMap<i64, i64>, x: i64) -> bool {
    // Find the interval with the largest start <= x
    if let Some((&_start, &end)) = map.range(..=x).next_back() {
        x <= end
    } else {
        false
    }
}

fn solve(path: &str) -> (i64, i64) {
    let (ranges, ids) = parse(path);
    let interval_map = build_interval_map(ranges);

    // Part 1: Count IDs in any interval
    let fresh = ids.iter().filter(|&&id| contains_point(&interval_map, id)).count() as i64;

    // Part 2: Total span of merged intervals
    let total: i64 = interval_map.iter().map(|(a, b)| b - a + 1).sum();

    (fresh, total)
}

fn main() {
    let start = Instant::now();
    let (p1, p2) = solve("input.txt");
    println!(
        "available_fresh={} total_fresh_ids={} elapsed_ms={:.3}",
        p1, p2, start.elapsed().as_secs_f64() * 1000.0
    );
}
